tsdb/importer/import.go

```diff
@@ -0,0 +1,305 @@
+// Copyright 2019 The Prometheus Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package importer
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"math"
+	"os"
+	"path/filepath"
+	"strings"
+	"time"
+
+	"github.com/go-kit/kit/log/level"
+
+	"github.com/go-kit/kit/log"
+	"github.com/otiai10/copy"
+	prom_labels "github.com/prometheus/prometheus/pkg/labels"
+	"github.com/prometheus/prometheus/pkg/textparse"
+	"github.com/prometheus/prometheus/tsdb"
+	tsdb_labels "github.com/prometheus/prometheus/tsdb/labels"
+)
+
+// Implementing the error interface to create a
+// constant, which cannot be overridden.
+// https://dave.cheney.net/2016/04/07/constant-errors
+type Error string
+
+func (e Error) Error() string {
+	return string(e)
+}
+
+// This error is thrown when we try to merge/add blocks to an existing TSDB instance,
+// and the new blocks have a time overlap with the current blocks.
+const OverlappingBlocksError = Error("blocks overlap with blocks currently in DB")
+
+// Duration of a block in milliseconds
+const BlockDuration = 2 * 60 * 60 * 1000
+
+type timestamp = int64
+
+type metricSample struct {
+	TimestampMs timestamp
+	Value       float64
+	Labels      tsdb_labels.Labels
+}
+
+// ImportFromFile imports data from a file formatted according to the Prometheus exposition format,
+// converts it into block(s), and places the newly created block(s) in the
+// TSDB DB directory, where it is treated like any other block.
+func ImportFromFile(filePath string, contentType string, dbPath string, skipTimestampCheck bool, logger log.Logger) error {
+	if logger == nil {
+		logger = log.NewNopLogger()
+	}
+
+	tmpDbDir, err := ioutil.TempDir("", "importer")
+	if err != nil {
+		return err
+	}
+	defer os.RemoveAll(tmpDbDir)
+
+	dbMint, dbMaxt, err := getDbTimeLimits(dbPath)
+	if err != nil {
+		return err
+	}
+
+	f, err := os.Open(filePath)
+	if err != nil {
+		return err
+	}
+	defer f.Close()
+
+	bytes, err := ioutil.ReadAll(f)
```

Comment:
this load the entyre file in memory. In general I would say that using mmap and reading it chunks would be much more memory efficient, but if/when needed can do this in a separate PR.

Reply:
Makes sense.

Reply:
File sizes may be massive, this needs streaming of some form.

Reply:
Yes, @krasi-georgiev suggested tackling that in a follow-up PR to address that potential issue.